spark/src/main/scala/org/apache/spark/sql/delta/commands/OptimizeTableCommand.scala

```diff
@@ -400,8 +393,93 @@ class OptimizeExecutor(
           bin.size > 1 || // bin has more than one file or
           bin.size == 1 && optimizeContext.reorg.nonEmpty || // always rewrite files during reorg
           isMultiDimClustering // multi-clustering
-        }.map(b => (partition, b))
+        }.map(b => Bin(partition, b))
+    }
+  }
+
+  /**
+   * Utility methods to group bins into batches for incremental progress.
+   *
+   * @param bins List of bins to group into batches
+   * @return Sequence of batches. Each batch contains one or more bins.
+   */
+  private def groupBinsIntoBatches(
+      bins: Seq[Bin],
+      batchSize: Long)
+  : Seq[Batch] = {
+    val batches = new ArrayBuffer[Batch]()
+
+    val currentBatch = new ArrayBuffer[Bin]()
+    var currentBatchSize = 0L
+
+    bins.foreach { bin =>
```

Comment:
We have similar bin packing logic in [BinPackingUtils](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/util/BinPackingUtils.scala), can we reuse it?

Reply:
Hmm yeah I think so. The only difference is it emits a bin before it reaches the maxBinSize, where I currently emit a bit right after I reach it, but I don't think it matters that much either way

Reply:
Replaced the function with that and just inlined it above